airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java

```diff
@@ -22,97 +24,106 @@
+          } catch (RuntimeException e) {
+            _exception = e;
+          }
+        }
+      }
     }
 
-    private synchronized C getOrCreate(ContainerFactory<C> factory) {
-      if (sharedContainer == null && containerCreationError == null) {
-        try {
-          create(imageName, factory, methodNames);
-        } catch (RuntimeException e) {
-          sharedContainer = null;
-          containerCreationError = e;
-        }
+    RuntimeException exception() {
+      if (_exception == null && _container == null) {
+        populate();
       }
-      if (containerCreationError != null) {
-        throw new RuntimeException(
-            "Error during container creation for imageName=" + imageName
-                + ", factory=" + factory.getClass().getName()
-                + ", methods=" + methodNames,
-            containerCreationError);
+      return _exception;
+    }
+
+    GenericContainer<?> container() {
+      if (_exception == null && _container == null) {
+        populate();
       }
-      return sharedContainer;
+      return _container;
     }
 
-    private void create(String imageName, ContainerFactory<C> factory, List<String> methodNames) {
+  }
+
+  private static final ConcurrentMap<ContainerKey, ContainerOrException> SHARED_CONTAINERS = new ConcurrentHashMap<>();
+
+  /**
+   * Creates a new, unshared testcontainer instance. This usually wraps the default constructor for
+   * the testcontainer type.
+   */
+  protected abstract C createNewContainer(DockerImageName imageName);
+
+  /**
+   * Returns a shared instance of the testcontainer.
+   */
+  public final C shared(String imageName, String... methods) {
+    List<String> methodList = methods == null ? Collections.emptyList() : Arrays.asList(methods);
+    DockerImageName dockerImageName = DockerImageName.parse(imageName);
+    final ContainerKey containerKey = new ContainerKey(getClass(), dockerImageName, methodList);
+    ContainerOrException containerOrError = SHARED_CONTAINERS.computeIfAbsent(containerKey, this::createContainerOrError);
+    if (containerOrError.exception() != null) {
+      throw containerOrError.exception();
+    }
+    return (C) containerOrError.container();
+  }
+
+  public final C exclusive(String imageName, String... methods) {
+    DockerImageName dockerImageName = DockerImageName.parse(imageName);
+    List<String> methodList = methods == null ? Collections.emptyList() : Arrays.asList(methods);
+    return (C) createContainerSupplier(dockerImageName, methodList).get();
+  }
+
+  private ContainerOrException createContainerOrError(ContainerKey containerKey) {
+    DockerImageName imageName = containerKey.imageName();
+    List<String> methodNames = containerKey.methods();
+    return new ContainerOrException(createContainerSupplier(imageName, methodNames));
+  }
+
+  private Supplier<GenericContainer<?>> createContainerSupplier(DockerImageName imageName, List<String> methodNames) {
```

Comment:
It might be more elegant to have this function's return value be a `GenericContainer<?>` instead, and have `createContainerOrError` end with `return new ContainerOrException(() -> createAndStartContainer(imageName, methodNames));` or something